components-api/src/main/scala/pl/touk/nussknacker/engine/api/definition/ParameterDeclaration.scala

```diff
@@ -0,0 +1,258 @@
+    val underlying = paramType.extractor
+    new ParameterExtractor[T#EXTRACTED_VALUE_TYPE] with ParameterCreatorWithNoDependency {
+      override def parameterName: ParameterName                         = underlying.parameterName
+      override def extractValue(params: Params): T#EXTRACTED_VALUE_TYPE = underlying.extractValue(params)
+      override def createParameter(): Parameter                         = modify(underlying.createBase)
+      override def createBase: Parameter                                = underlying.createBase
+    }
+  }
+
+  def withAdvancedCreator[DEPENDENCY](
+      create: DEPENDENCY => Parameter => Parameter
+  ): ParameterExtractor[T#EXTRACTED_VALUE_TYPE] with ParameterCreator[DEPENDENCY] = {
+    val underlying = paramType.extractor
+    new ParameterExtractor[T#EXTRACTED_VALUE_TYPE] with ParameterCreator[DEPENDENCY] {
+      override def parameterName: ParameterName                         = underlying.parameterName
+      override def extractValue(params: Params): T#EXTRACTED_VALUE_TYPE = underlying.extractValue(params)
+      override def createParameter: DEPENDENCY => Parameter = dependency => create(dependency)(underlying.createBase)
+      override def createBase: Parameter                    = underlying.createBase
+    }
+  }
+
+}
+
+object ParameterDeclarationBuilder {
+
+  sealed abstract class ParamType {
+    type EXTRACTED_VALUE_TYPE
+
+    def extractor: ParameterExtractor[EXTRACTED_VALUE_TYPE]
+  }
+
+  object ParamType {
+
+    final case class Mandatory[T: TypeTag: NotNothing](parameterName: ParameterName) extends ParamType {
+      override type EXTRACTED_VALUE_TYPE = T
+      override lazy val extractor: ParameterExtractor[EXTRACTED_VALUE_TYPE] =
+        new MandatoryParamExtractor[T](parameterName)
+    }
+
+    final case class LazyMandatory[T <: AnyRef: TypeTag: NotNothing](parameterName: ParameterName) extends ParamType {
+      override type EXTRACTED_VALUE_TYPE = LazyParameter[T]
+      override lazy val extractor: ParameterExtractor[EXTRACTED_VALUE_TYPE] =
+        new MandatoryLazyParamExtractor[T](parameterName)
+    }
+
+    final case class BranchMandatory[T: TypeTag: NotNothing](parameterName: ParameterName) extends ParamType {
+      override type EXTRACTED_VALUE_TYPE = Map[String, T]
+      override lazy val extractor: ParameterExtractor[Map[String, T]] = new MandatoryBranchParamExtractor(parameterName)
+    }
+
+    final case class BranchLazyMandatory[T <: AnyRef: TypeTag: NotNothing](parameterName: ParameterName)
+        extends ParamType {
+      override type EXTRACTED_VALUE_TYPE = Map[String, LazyParameter[T]]
+      override lazy val extractor: ParameterExtractor[Map[String, LazyParameter[T]]] =
+        new MandatoryBranchLazyParamExtractor[T](parameterName)
+    }
+
+    final case class Optional[T: TypeTag: NotNothing](parameterName: ParameterName) extends ParamType {
+      override type EXTRACTED_VALUE_TYPE = Option[T]
```

Comment:
In `pl.touk.nussknacker.engine.api.definition.Parameter` there are both:
```
    scalaOptionParameter: Boolean,
    javaOptionalParameter: Boolean,
```
so do we need another class to cover Java Optional or is this automatically converted to Scala Option?

Reply:
IMO it will be covered.
But let's @arkadius confirm it

Reply:
This `scalaOptionParameter` and `javaOptionalParameter` are for `@MethodToInvoke` based components and it is another thing that IMO we should phase out. The approach provided by @mk-software-pl is better because it better handles combinations like mandatory-but-nullable-and-not-CompileTimeEvaluableValue and ;-) similar edge cases